appyx-components/experimental/puzzle15/common/src/commonMain/kotlin/com/bumble/appyx/components/experimental/puzzle15/ui/Puzzle15Ui.kt

```diff
@@ -105,17 +109,35 @@ fun Puzzle15Ui(
                                 .then(
                                     elementUiModel.modifier
                                         .background(color = Color.White)
-                                ).pointerInput(elementUiModel.element.id) {
-                                    detectDragGestures(
-                                        onDrag = { change, dragAmount ->
-                                            change.consume()
-                                            puzzle15.onDrag(dragAmount, this)
-                                        },
-                                        onDragEnd = {
-                                            AppyxLogger.d("drag", "end")
-                                            puzzle15.onDragEnd()
+                                )
+                                .composed {
+                                    var isDragging by remember { mutableStateOf(false) }
+                                    DisposableEffect(Unit) {
+                                        onDispose {
+                                            if (isDragging) {
+                                                puzzle15.onDragEnd()
+                                            }
                                         }
-                                    )
+                                    }
+                                    pointerInput(elementUiModel.element.id) {
+                                        detectDragGestures(
+                                            onDragStart = {
+                                                isDragging = true
+                                            },
+                                            onDragCancel = {
+                                                isDragging = false
+                                            },
+                                            onDrag = { change, dragAmount ->
+                                                change.consume()
+                                                isDragging = true
+                                                puzzle15.onDrag(dragAmount, this)
+                                            },
+                                            onDragEnd = {
+                                                isDragging = false
+                                                puzzle15.onDragEnd()
+                                            }
+                                        )
+                                    }
```

Comment:
Can we promote it and provide a more generic API for clients? Maybe we can name it `appyxPointerInput` or `appyxDragGesture`.

Reply:
my question is rather why we don't use DraggableChildren here?